File: api/src/modules/users/service.ts
import { randomUUID } from "node:crypto";

import { eq, inArray, sql } from "drizzle-orm";

import { db } from "../../db/client.js";
import { geographies, userRoles, users, userGeographyScopes } from "../../db/schema.js";
import { chartRoles, type ChartRole, type CurrentUserContext } from "../auth/types.js";
import {
  disableKeycloakUser,
  upsertKeycloakUser,
  type IdentityUser,
  type IdentityUserInput,
} from "./keycloakIdentity.js";
import { UserError } from "./errors.js";
import type { CreateUserInput, UserRecord } from "./types.js";

type UserIdentityProvider = {
  upsertUser(input: IdentityUserInput): Promise<IdentityUser>;
  disableUser(userId: string): Promise<void>;
};

type GeographyRow = typeof geographies.$inferSelect;
type UserRow = typeof users.$inferSelect;

export interface UserService {
  listUsers(context: CurrentUserContext): Promise<UserRecord[]>;
  createUser(input: CreateUserInput, context: CurrentUserContext): Promise<UserRecord>;
  disableUser(userId: string, context: CurrentUserContext): Promise<UserRecord>;
}

const defaultIdentityProvider: UserIdentityProvider = {
  upsertUser: upsertKeycloakUser,
  disableUser: disableKeycloakUser,
};

export function createUserService(
  identityProvider: UserIdentityProvider = defaultIdentityProvider,
): UserService {
  return {
    async listUsers(context) {
      assertCanManageUsers(context);

      return readUsers();
    },

    async createUser(input, context) {
      assertCanManageUsers(context);

      const normalized = normalizeCreateUserInput(input);
      const geographyRows = await findGeographiesById(normalized.geographyIds);

      if (geographyRows.length !== normalized.geographyIds.length) {
        throw new UserError("USER_GEOGRAPHY_INVALID", 400);
      }

      const identityUser = await identityProvider.upsertUser({
        name: normalized.name,
        email: normalized.email,
        phone: normalized.phone,
        username: normalized.username,
        password: normalized.password,
        roles: normalized.roles,
        groupPaths: geographyRows.map((geography) => geography.path),
      });

      await persistUserProjection({
        userId: identityUser.userId,
        username: identityUser.username,
        email: identityUser.email,
        phone: normalized.phone,
        displayName: normalized.name,
        roles: normalized.roles,
        geographies: geographyRows,
        createdByUserId: context.userId,
        source: "admin",
      });

      const created = await readUser(identityUser.userId);

      if (!created) {
        throw new UserError("USER_NOT_FOUND", 404);
      }

      return created;
    },

    async disableUser(userId, context) {
      assertCanManageUsers(context);

      if (userId === context.userId) {
        throw new UserError("USER_CANNOT_DISABLE_SELF", 400);
      }

      const existing = await readUser(userId);

      if (!existing) {
        throw new UserError("USER_NOT_FOUND", 404);
      }

      await identityProvider.disableUser(userId);

      await db
        .update(users)
        .set({
          status: "disabled",
          updatedAt: sql`now()`,
        })
        .where(eq(users.id, userId));

      return {
        ...existing,
        status: "disabled",
      };
    },
  };
}

export async function persistUserProjection(input: {
  userId: string;
  username: string;
  email?: string;
  phone?: string;
  displayName: string;
  roles: ChartRole[];
  geographies: GeographyRow[];
  createdByUserId?: string;
  source: string;
}) {
  await db.transaction(async (tx) => {
    await tx
      .insert(users)
      .values({
        id: input.userId,
        username: input.username,
        email: input.email ?? null,
        phone: input.phone ?? null,
        displayName: input.displayName,
        status: "active",
        createdByUserId: input.createdByUserId,
        lastSeenAt: sql`now()`,
      })
      .onConflictDoUpdate({
        target: users.id,
        set: {
          username: sql`excluded.username`,
          email: sql`excluded.email`,
          phone: sql`excluded.phone`,
          displayName: sql`excluded.display_name`,
          status: "active",
          lastSeenAt: sql`now()`,
          updatedAt: sql`now()`,
        },
      });

    await tx.delete(userRoles).where(eq(userRoles.userId, input.userId));

    if (input.roles.length > 0) {
      await tx.insert(userRoles).values(
        input.roles.map((role) => ({
          userId: input.userId,
          role,
          source: input.source,
          syncedAt: sql`now()`,
        })),
      );
    }

    await tx
      .delete(userGeographyScopes)
      .where(eq(userGeographyScopes.userId, input.userId));

    if (input.geographies.length > 0) {
      await tx.insert(userGeographyScopes).values(
        input.geographies.map((geography) => ({
          id: `user-geo-${randomUUID()}`,
          userId: input.userId,
          geographyId: geography.id,
          source: input.source,
          externalGroupPath: geography.path,
        })),
      );
    }
  });
}

async function readUsers() {
  const rows = await db.select().from(users);
  const roles = await db.select().from(userRoles);
  const scopes = await readAllUserGeographyScopes();

  return rows
    .map((row) =>
      mapUserRecord(
        row,
        roles
          .filter((role) => role.userId === row.id)
          .map((role) => role.role as ChartRole),
        scopes.get(row.id) ?? [],
      ),
    )
    .sort((first, second) => first.displayName.localeCompare(second.displayName));
}

async function readUser(userId: string) {
  const rows = await db.select().from(users).where(eq(users.id, userId)).limit(1);
  const row = rows[0];

  if (!row) {
    return undefined;
  }

  const roleRows = await db
    .select()
    .from(userRoles)
    .where(eq(userRoles.userId, userId));
  const scopes = await readAllUserGeographyScopes(userId);

  return mapUserRecord(
    row,
    roleRows.map((role) => role.role as ChartRole),
    scopes.get(userId) ?? [],
  );
}

async function readAllUserGeographyScopes(userId?: string) {
  const rows = await db
    .select({
      userId: userGeographyScopes.userId,
      geographyId: geographies.id,
      path: geographies.path,
      name: geographies.name,
      levelLabel: geographies.levelLabel,
    })
    .from(userGeographyScopes)
    .innerJoin(geographies, eq(userGeographyScopes.geographyId, geographies.id))
    .where(userId ? eq(userGeographyScopes.userId, userId) : undefined);
  const scopesByUser = new Map<string, UserRecord["geographyScopes"]>();

  for (const row of rows) {
    const scopes = scopesByUser.get(row.userId) ?? [];

    scopes.push({
      geographyId: row.geographyId,
      path: row.path,
      name: row.name,
      levelLabel: row.levelLabel,
    });
    scopesByUser.set(row.userId, scopes);
  }

  return scopesByUser;
}

function mapUserRecord(
  row: UserRow,
  roles: ChartRole[],
  geographyScopes: UserRecord["geographyScopes"],
): UserRecord {
  return {
    userId: row.id,
    username: row.username,
    email: row.email ?? undefined,
    phone: row.phone ?? undefined,
    displayName: row.displayName,
    status: row.status,
    roles,
    geographyScopes,
  };
}

async function findGeographiesById(geographyIds: string[]) {
  if (geographyIds.length === 0) {
    return [];
  }

  return db.select().from(geographies).where(inArray(geographies.id, geographyIds));
}

function normalizeCreateUserInput(input: CreateUserInput): CreateUserInput {
  const name = input.name.trim();
  const email = input.email.trim().toLowerCase();
  const phone = input.phone?.trim();
  const username = input.username.trim().toLowerCase();
  const roles = uniqueValues(input.roles);
  const geographyIds = uniqueValues(input.geographyIds);

  if (!name) {
    throw new UserError("USER_NAME_REQUIRED", 400);
  }

  if (!email) {
    throw new UserError("USER_EMAIL_REQUIRED", 400);
  }

  if (!username) {
    throw new UserError("USER_USERNAME_REQUIRED", 400);
  }

  if (input.password.length < 8) {
    throw new UserError("USER_PASSWORD_REQUIRED", 400);
  }

  if (roles.length === 0) {
    throw new UserError("USER_ROLE_REQUIRED", 400);
  }

  if (roles.some((role) => !chartRoles.includes(role))) {
    throw new UserError("USER_ROLE_INVALID", 400);
  }

  if (geographyIds.length === 0) {
    throw new UserError("USER_GEOGRAPHY_REQUIRED", 400);
  }

  return {
    name,
    email,
    phone,
    username,
    password: input.password,
    roles,
    geographyIds,
  };
}

function assertCanManageUsers(context: CurrentUserContext) {
  if (!context.roles.includes("chart_admin")) {
    throw new UserError("USER_FORBIDDEN", 403);
  }
}

function uniqueValues<T extends string>(values: T[]) {
  return [...new Set(values.map((value) => value.trim()).filter(Boolean))] as T[];
}

File: backend/chart/users/service.py
from __future__ import annotations

import logging
import uuid

from sqlalchemy import delete, select

from chart.auth.schemas import CurrentUserContext
from chart.auth.service import chart_roles
from chart.identity import IdentityError, disable_user as disable_identity_user
from chart.identity import upsert_user
from chart.shared.db.models import (
    AppGeography,
    AppUser,
    UserGeographyScopeRecord,
    UserRoleRecord,
)
from chart.shared.db.session import get_session_factory

from .schemas import CreateUserInput, UserGeographyScope, UserResponse

logger = logging.getLogger(__name__)


class UserServiceError(ValueError):
    def __init__(self, code: str, status_code: int) -> None:
        super().__init__(code)
        self.code = code
        self.status_code = status_code


def list_users(user: CurrentUserContext) -> list[UserResponse]:
    _require_admin(user)
    with get_session_factory()() as session:
        users = list(session.scalars(select(AppUser)))
        user_ids = [item.id for item in users]
        roles: dict[str, list[str]] = {user_id: [] for user_id in user_ids}
        for user_id, role in session.execute(
            select(UserRoleRecord.user_id, UserRoleRecord.role).where(
                UserRoleRecord.user_id.in_(user_ids)
            )
        ):
            roles[user_id].append(role)
        places: dict[str, list[AppGeography]] = {user_id: [] for user_id in user_ids}
        for user_id, place in session.execute(
            select(UserGeographyScopeRecord.user_id, AppGeography)
            .join(
                AppGeography,
                AppGeography.id == UserGeographyScopeRecord.geography_id,
            )
            .where(UserGeographyScopeRecord.user_id.in_(user_ids))
        ):
            places[user_id].append(place)
        return [_user_response(item, roles[item.id], places[item.id]) for item in users]


def create_user(input_data: CreateUserInput, actor: CurrentUserContext) -> UserResponse:
    _require_admin(actor)
    roles = list(dict.fromkeys(input_data.roles))
    if any(role not in chart_roles for role in roles):
        raise UserServiceError("USER_ROLE_INVALID", 400)
    geography_ids = list(dict.fromkeys(input_data.geographyIds))
    with get_session_factory()() as session:
        places = list(
            session.scalars(
                select(AppGeography).where(AppGeography.id.in_(geography_ids))
            )
        )
    if len(places) != len(geography_ids):
        raise UserServiceError("USER_GEOGRAPHY_INVALID", 400)
    try:
        identity = upsert_user(
            name=input_data.name.strip(),
            email=input_data.email.strip().lower(),
            username=input_data.username.strip().lower(),
            password=input_data.password,
            roles=roles,
            group_paths=[place.path for place in places],
        )
    except IdentityError as error:
        raise UserServiceError(error.code, error.status_code) from error

    persisted = False
    try:
        with get_session_factory()() as session:
            record = session.get(AppUser, identity.user_id)
            if record is None:
                record = AppUser(
                    id=identity.user_id,
                    username=identity.username,
                    display_name=input_data.name.strip(),
                )
                session.add(record)
            record.email = identity.email
            record.phone = input_data.phone.strip() if input_data.phone else None
            record.display_name = input_data.name.strip()
            record.status = "active"
            record.created_by_user_id = actor.user_id
            session.execute(
                delete(UserRoleRecord).where(UserRoleRecord.user_id == record.id)
            )
            session.execute(
                delete(UserGeographyScopeRecord).where(
                    UserGeographyScopeRecord.user_id == record.id
                )
            )
            for role in roles:
                session.add(
                    UserRoleRecord(user_id=record.id, role=role, source="admin")
                )
            for place in places:
                session.add(
                    UserGeographyScopeRecord(
                        id=f"user-geo-{uuid.uuid4()}",
                        user_id=record.id,
                        geography_id=place.id,
                        source="admin",
                        external_group_path=place.path,
                    )
                )
            session.commit()
            persisted = True
            return _response(session, record)
    except Exception as error:
        if identity.created and not persisted:
            try:
                disable_identity_user(identity.user_id)
            except IdentityError:
                logger.exception(
                    "Failed to disable orphaned identity %s", identity.user_id
                )
        raise UserServiceError("USER_PERSIST_FAILED", 500) from error


def disable_user(user_id: str, actor: CurrentUserContext) -> UserResponse:
    _require_admin(actor)
    if user_id == actor.user_id:
        raise UserServiceError("USER_CANNOT_DISABLE_SELF", 400)
    with get_session_factory()() as session:
        record = session.get(AppUser, user_id)
        if record is None:
            raise UserServiceError("USER_NOT_FOUND", 404)
    try:
        disable_identity_user(user_id)
    except IdentityError as error:
        raise UserServiceError(error.code, error.status_code) from error
    with get_session_factory()() as session:
        record = session.get(AppUser, user_id)
        assert record is not None
        record.status = "disabled"
        session.commit()
        return _response(session, record)


def _response(session, user: AppUser) -> UserResponse:
    roles = list(
        session.scalars(
            select(UserRoleRecord.role).where(UserRoleRecord.user_id == user.id)
        )
    )
    places = session.execute(
        select(AppGeography)
        .join(
            UserGeographyScopeRecord,
            UserGeographyScopeRecord.geography_id == AppGeography.id,
        )
        .where(UserGeographyScopeRecord.user_id == user.id)
    ).scalars()
    return _user_response(user, roles, list(places))


def _user_response(
    user: AppUser,
    roles: list[str],
    places: list[AppGeography],
) -> UserResponse:
    return UserResponse(
        userId=user.id,
        username=user.username,
        email=user.email,
        phone=user.phone,
        displayName=user.display_name,
        status=user.status,
        roles=roles,
        geographyScopes=[
            UserGeographyScope(
                geographyId=place.id,
                path=place.path,
                name=place.name,
                levelLabel=place.level_label,
            )
            for place in places
        ],
    )


def _require_admin(user: CurrentUserContext) -> None:
    if "chart_admin" not in user.roles:
        raise UserServiceError("USER_FORBIDDEN", 403)
